src/main/java/com/routeme/service/SearchServiceImpl.java

```diff
@@ -80,51 +116,189 @@ private SearchResponseDTO convertRouteDTOsToSearchResponseDTO(List<RouteDTO> rou
+            stepDTO = new NonTransitStepDTO(nonTransitStep.getTransportationMode(), nonTransitStep.getHtmlIntruction(),
+                    nonTransitStep.getDistance());
+        }
+        return stepDTO;
+    }
+}
+
+class RoutePreferenceSorter implements Comparator<RouteDTO> {
+    ArrayList<String> travelModePreference = new ArrayList<String>();
+    ArrayList<String> routeTypePreference = new ArrayList<String>();
+
+    public RoutePreferenceSorter(ArrayList<String> travelModePreference, ArrayList<String> routeTypePreference) {
+        this.travelModePreference = travelModePreference;
+        this.routeTypePreference = routeTypePreference;
+    }
+
+    @Override
+    public int compare(RouteDTO routeDTO1, RouteDTO routeDTO2) {
+        int likedRouteScore = compareWithRespectToLikes(routeDTO1, routeDTO2);
+        if (likedRouteScore == 0) {
+            int travelModePrefScore = compareWithRespectToTravelModePreference(routeDTO1, routeDTO2);
+            if (travelModePrefScore == 0) {
+                return compareWithRespectToRouteTypePreferences(routeDTO1, routeDTO2);
+            } else {
+                return travelModePrefScore;
+            }
+        } else {
+            return likedRouteScore;
+        }
+    }
+
+    private int compareWithRespectToLikes(RouteDTO routeDTO1, RouteDTO routeDTO2) {
+        if (routeDTO1.isLiked() == routeDTO2.isLiked()) {
+            return 0;
+        } else {
+            if (routeDTO1.isLiked()) {
+                return -1;
+            } else {
+                return 1;
+            }
+        }
+    }
+
+    private int compareWithRespectToTravelModePreference(RouteDTO routeDTO1, RouteDTO routeDTO2) {
+        ArrayList<String> routeDTO1Modes = routeDTO1.getTransportationModes();
+        ArrayList<String> routeDTO2Modes = routeDTO2.getTransportationModes();
+        int totalTransportationModesScore = 0;
+        for (String transportationMode1 : routeDTO1Modes) {
+            int transportationMode1Score = travelModePreference.indexOf(transportationMode1);
+            for (String transportationMode2 : routeDTO2Modes) {
+                int transportationMode2Score = travelModePreference.indexOf(transportationMode2);
+                totalTransportationModesScore += transportationMode1Score - transportationMode2Score;
+            }
+        }
+        return totalTransportationModesScore;
+    }
+
+    private int compareWithRespectToRouteTypePreferences(RouteDTO routeDTO1, RouteDTO routeDTO2) {
+        int leastTimeScore = (int) (routeDTO1.getDurationInSeconds() - routeDTO2.getDurationInSeconds());
+        int leastNumberOfChangesScore = routeDTO1.getNumberOfChanges() - routeDTO2.getNumberOfChanges();
+        int soonestScore = (int) (routeDTO2.getDepartureDateTimeInMillis() - routeDTO1.getDepartureDateTimeInMillis());
+        int highPrefScore = 0;
+        int lowPrefScore = 0;
+        switch (routeTypePreference.get(0)) {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Switch statements should have a default label](https://www.codacy.com/app/heshamhamdymassoud/RouteMe-API/file/2647727881/issues/source?bid=3287636&fileBranchId=3323263#l281)